doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,310 @@
+limit. For further detail see the implementation function \ref
+goto_partial_inline(goto_modelt &, message_handlert &, unsigned, bool)
+
+<em>Predecessor pass is \ref precondition-transform.</em>
+
+
+\subsection assertions-transform Transform Assertions Assumptions
+ 
+This pass removes user provided assertions and assumptions when user has opted
+to do so. Note that this pass removes skip instructions if any other changes
+are made. The implementation of this pass is called via the \ref
+transform_assertions_assumptions(const optionst &, goto_modelt &) function.
+
+<em>Predecessor pass is \ref check-c-transform.</em>
+
+
+\subsection string-abstraction-transform String Abstraction
+
+This (optional) transformation pass adds tracking of length of C strings. It is
+switched on by the `--string-abstraction` command line option. This changes made
+by this pass are documented as part of the documentation for the \ref
+string_abstractiont class. The implementation of this pass is called via the
+\ref string_abstraction(goto_modelt &, message_handlert &) function.
+
+<em>Predecessor pass is \ref floats-transform.</em>
+
+
+\subsection nondet-transform Add Non-Deterministic Initialisation of Global Scoped Variables
+
+This transformation adds non-deterministic initialisation of global scoped
+variables including static variables. For details see
+`src/goto-instrument/nondet_static.h`. The initialisation code is added to the
+`CPROVER_initialize` function in the goto model. The implementation of this pass
+is called via the \ref nondet_static(goto_modelt &) function.
+
+<em>Predecessor pass is \ref update-transform.</em>
+
+
+\subsection unused-functions-transform Remove Unused Functions
+
+This pass removes unused functions from the goto model. In practice this builds
+a collection of all the functions that are potentially called, and then removes
+any function not in this collection.
+
+The implementation of this pass is called via the \ref
+remove_unused_functions(goto_modelt &, message_handlert &) function.
+
+<em>Predecessor pass is \ref failed-symbols-transform.</em>
+
+
+\subsection coverage-transform Add Coverage Goals
+
+This transformation adds coverage goals instrumentation and is only applied if
+the `--cover` option has been specified. The implementation of this pass is
+called via the \ref instrument_cover_goals(const cover_configt &, goto_modelt &, message_handlert &)
+function.
+
+<em>Predecessor pass is \ref remove-skip-transform .</em>
+
+
+\subsection slicing-transforms Slicing
+
+This includes several slicing passes as specified by various slicing command
+line options. The reachability slicer is enabled by the `--reachability-slice`
+command line option. The implementation of this pass is called via the \ref
+reachability_slicer(goto_modelt &, message_handlert &) function. The full slicer
+is enabled by the `--full-slice` command line option. The implementation of this
```

Comment:
Should we add some links to the documents that mentioned some of these options were unsound? https://github.com/diffblue/cbmc/pull/6479/files